Doctype ViewHelper Factory

rieschl opened the pull request:
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | yes

### Description

This PR adds a factory for the `Doctype` view helper, to be able to configure the used doctype via configuration.
The factory is registered with the default factories in the `HelperPluginManager`.
Also, a test checks if the default doctype is HTML4_LOOSE.

fixes #50

rieschl committed:
feat: add a factory for Doctype ViewHelper

Register the Factory in the HelperPluginManager

Signed-off-by: Thomas Rieschl <thomas@trinet.at>

rieschl commented on composer.json:
```diff
@@ -19,6 +19,9 @@
     "extra": {
+        "branch-alias": {
+            "dev-doctype-factory": "2.12.x-dev"
+        }
```
That workaround fixes the [failure in Travis](https://travis-ci.com/github/laminas/laminas-view/jobs/399039557#L236). But that's specific to the current feature branch. I have no idea how to fix that.

rieschl replied:
solved through  [`COMPOSER_ROOT_VERSION`](https://github.com/laminas/laminas-view/pull/59/files#diff-6ac3f79fc25d95cd1e3d51da53a4b21b939437392578a35ae8cd6d5366ca5485R10) env variable

rieschl commented on docs/book/helpers/doctype.md:
```diff
@@ -26,39 +26,58 @@ You can also specify a custom doctype as long as it is well-formed.
     /* ... */
     'view_manager' => [
-        'doctype' => 'html5',
+        'doctype' => 'HTML5',
```
When writing the docs I noticed that the doctype has to be in uppercase, even in `laminas-mvc`.

@weierophinney @froschdesign Is it okay if I change that here or do you want me to change that in a separate PR? Or maybe even change `Doctype` itself to accept lowercase?

froschdesign replied:
@rieschl
> Is it okay if I change that here…

Using a separate commit on this pull request is enough. Btw. good catch! 👍 

> Or maybe even change Doctype itself to accept lowercase?

Good idea! Create an issue report or a pull request – both are welcome.

rieschl replied:
Thanks for the quick reply!


> @rieschl
> > Is it okay if I change that here…
> 
> Using a separate commit on this pull request is enough.

Will do!

> Btw. good catch! 👍 

Thanks. finding and fixing bugs, code reviewing and keeping code quality (tools) up to date are my main tasks in my biggest project, so I kind of have an eye for things like that 🙂


> 
> > Or maybe even change Doctype itself to accept lowercase?
> 
> Good idea! Create an issue report or a pull request – both are welcome.
> 

okay I'll come up with a PR, but probably not today anymore.

rieschl replied:
> 
> Using a separate commit on this pull request is enough.

Done :)

rieschl commented:
Btw. I also tried it directly in a Mezzio project. Works as expected.

froschdesign commented:
@rieschl 
I think we should wait for the PHP 8 support and the remove of laminas-servicemanager with version 2.

rieschl commented:
> 
> 
> @rieschl
> I think we should wait for the PHP 8 support and the remove of laminas-servicemanager with version 2.

Okay, should I do that? PHP8 and dropping servicemanager v2 in one PR?

froschdesign commented:
@rieschl 
> PHP8 and dropping servicemanager v2 in one PR?

Yes, version 2 of laminas-servicemanager is no longer supported and does not support PHP 8.
Please have a look at #57 and #58.

Thanks in advance! 👍

rieschl commented:
> 
> 
> @rieschl
> 
> > PHP8 and dropping servicemanager v2 in one PR?
> 
> Yes, version 2 of laminas-servicemanager is no longer supported and does not support PHP 8.
> Please have a look at #57 and #58.
> 
> Thanks in advance! 👍

Ah, I didn't notice that there is already a PR for PHP8. Thanks!

froschdesign commented on docs/book/helpers/doctype.md:
```diff
@@ -26,39 +26,58 @@ You can also specify a custom doctype as long as it is well-formed.
-## Basic Usage
+## Configuration and Basic Usage
 
 You may specify the doctype at any time. However, helpers that depend on the
-doctype for their output will recognize it only after you have set it, so the
-easiest approach is to specify it in your bootstrap:
+doctype for their output will recognize it only after you have set it.
+There are multiple ways to set the doctype:
```
This section should be reverted because "Basic Usage" is the first correct entry. The previous version was okay.

rieschl replied:
Done! I borrowed the headline from the `asset` documentation 🙂 

Maybe you should come up with a Documentation guide, like the [Maintainers guide](https://github.com/laminas/technical-steering-committee/blob/main/MAINTAINERS.md) 😉

froschdesign commented on docs/book/helpers/doctype.md:
```diff
@@ -26,39 +26,58 @@ You can also specify a custom doctype as long as it is well-formed.
 
-Within an application based off the [laminas-mvc-skeleton](https://github.com/laminas/laminas-mvc-skeleton),
-you can specify the doctype via configuration:
+### `laminas-mvc` configuration
+
+An application based off the [laminas-mvc-skeleton](https://github.com/laminas/laminas-mvc-skeleton)
+comes with its own factory in the `ViewHelperManager` and can be configured like this:
```
The headline should be "Usage in a laminas-mvc Application" and the second level should be used.

froschdesign replied:
The link to the "ViewManager" from the end of the page can be added here. [The last headline and paragraph](https://docs.laminas.dev/laminas-view/helpers/doctype/#laminas-mvc-view-manager) of the current page can be removed then.

rieschl replied:
done!

froschdesign commented on docs/book/helpers/doctype.md:
````diff
@@ -26,39 +26,58 @@ You can also specify a custom doctype as long as it is well-formed.
-```
-
-And then print it out on top of your layout script:
+`\Laminas\View\Helper\Service\DoctypeFactory` checks the application configuration, making it possible to
+define the doctype through your configuration (e.g. a `ConfigProvider`):
 
 ```php
-<?php echo $this->doctype() ?>
+return [
+    'view_helper_config' => [
+        'doctype' => 'HTML5',
+    ],
+];
````
The headline should be "Usage in a Mezzio Application" and the second should level be used. The entire section should describe the usage in a Mezzio application.

rieschl replied:
done

rieschl committed:
docs: update Doctype documentation

Signed-off-by: Thomas Rieschl <thomas@trinet.at>

rieschl commented:
@froschdesign , thank you for your feedback! When the [PHP8 PR](https://github.com/laminas/laminas-view/pull/58) is merged, I'll update this PR and remove ServiceManager v2 stuff.

froschdesign commented on docs/book/helpers/doctype.md:
````diff
@@ -45,15 +45,30 @@ And then print it out on top of your layout script:
+];
+```
+
+## Laminas MVC View Manager
````
```suggestion
## Usage in a laminas-mvc Application
```
We already use this wording in other places.

rieschl replied:
argh damn. sorry, I already had it and then pasted the heading from the last paragraph over it 🤦‍♂️

froschdesign commented on docs/book/helpers/doctype.md:
```diff
@@ -45,15 +45,30 @@ And then print it out on top of your layout script:
+
+## Usage in a laminas-mvc Application
+
+If you're running a [LaminasMvc](https://github.com/laminas/laminas-mvc-skeleton) application, you should specify doctype via the
```
```suggestion
If you're running a [laminas-mvc](https://docs.laminas.dev/laminas-mvc/) application, you should specify doctype via the
```

boesing commented:
@rieschl I've rebased against latest 2.12.x (I've merged #58). Please apply your changes regarding SM v2 so we can release 2.12.0 after this 👍

rieschl committed:
qa: add .phpunit.result.cache to gitignore

Signed-off-by: Thomas Rieschl <thomas@trinet.at>

rieschl commented:
> 
> 
> @rieschl I've rebased against latest 2.12.x (I've merged #58). Please apply your changes regarding SM v2 so we can release 2.12.0 after this 👍

Thanks, @boesing ! I removed the v2 compatibility
I also added `.phpunit.result.cache` to the gitignore. It seems to have been forgotten.

froschdesign commented on docs/book/helpers/doctype.md:
```diff
@@ -45,15 +45,30 @@ And then print it out on top of your layout script:
     /* ... */
     'view_manager' => [
-        'doctype' => 'html5',
+        'doctype' => 'HTML5',
```
```suggestion
        'doctype' => Laminas\View\Helper\Doctype::HTML5,
```

froschdesign commented on docs/book/helpers/doctype.md:
```diff
@@ -45,15 +45,30 @@ And then print it out on top of your layout script:
-you can specify the doctype via configuration:
+## Usage in a Mezzio Application
+
+`\Laminas\View\Helper\Service\DoctypeFactory` checks the application configuration, making it possible to
+define the doctype through your configuration (e.g. a `ConfigProvider`):
```
```suggestion
The factory `Laminas\View\Helper\Service\DoctypeFactory` checks the application configuration, making it possible to define the doctype through your configuration, e.g. `config/autoload/mezzio.global.php`:
```

froschdesign commented on docs/book/helpers/doctype.md:
````diff
@@ -45,15 +45,30 @@ And then print it out on top of your layout script:
+```php
+return [
+    'view_helper_config' => [
+        'doctype' => 'HTML5',
````
```suggestion
        'doctype' => Laminas\View\Helper\Doctype::HTML5,
```

froschdesign reviewed (changes requested):
In a Mezzio application the global configuration can be used and also the class constants of `Laminas\View\Helper\Doctype`.

Thanks in advance!

rieschl commented:
> In a Mezzio application the global configuration can be used and also the class constants of `Laminas\View\Helper\Doctype`.


Thanks, again 😄 
I updated the docs

froschdesign commented on docs/book/helpers/doctype.md:
````diff
@@ -45,15 +45,32 @@ And then print it out on top of your layout script:
+or a `ConfigProvider.php` in a module.
+
+```php
+// config/autoload/mezzio.global.php
````
Please add the filename before the code example because there are always two parts for this information: what kind of file is this and what is the name or path.

Examples are:

> Create a file for the module configuration called `module.config.php` under `laminas-mvc-tutorial/module/Album/config/`:

> …in a factory, e.g. `src/Album/Controller/AlbumControllerFactory.php`:

> …in a module configuration file, e.g. `module/Album/config/module.config.php`:

In a code comment this would be too long and misplaced.

Thanks in advance!

_(@settermjd and I are still working on documentation writer's guide and we hope to publish it soon.)_

rieschl replied:
done, thanks!

I'm looking forward to a writer's guide! Then hopefully I'll get it right one the first try 🙂

froschdesign commented on docs/book/helpers/doctype.md:
```diff
@@ -45,15 +45,35 @@ And then print it out on top of your layout script:
+If you're running a [laminas-mvc](https://docs.laminas.dev/laminas-mvc/) application, you should specify doctype via the
+[ViewManager](https://docs.laminas.dev/laminas-mvc/services/#viewmanager) service.
+
+Add the following lines to your `module/Application/config/module.config.php` file to set the `Doctype` to HTML5:
```
```suggestion
Add the following lines to your `config/autoload/global.php` file to set the `Doctype` to HTML5:
```

This configuration is related to the entire application and not to a module. The same is already done in the previous section for Mezzio. (This is also wrong in the [skeleton application](https://github.com/laminas/laminas-mvc-skeleton/blob/12ff936b69caf63797b9c06a9f3a5d95482be913/module/Application/config/module.config.php#L48-L50).)

rieschl replied:
Thanks! Yes, I was always wondering why global stuff was in the `module` config. Do you want me to submit a PR against laminas-mvc-skeleton to fix that?

froschdesign commented on test/Helper/Service/DoctypeFactoryTest.php:
```diff
@@ -0,0 +1,69 @@
+
+    public function testServiceIsCreatedOnInvocation(): void
+    {
+        $container = $this->createContainer();
```
A mock object should be used here which is build by PHPUnit. The interface of the container is relevant here, not the concrete implementation like the service manager.
This saves us work in the future.

Thanks in advance!

rieschl replied:
> A mock object should be used here which is build by PHPUnit. The interface of the container is relevant here, not the concrete implementation like the service manager.
> This saves us work in the future.

Really? I almost never use mocks because of the [headache they caused](https://thephp.cc/dates/2020/06/php-fwdays-special-edition/avoiding-mockageddon) me in the past and always use real stuff, if possible. If the ServiceManager or Container changes, the test _should_ break because the Factory won't work in a real application anymore.
But if Mocks are the Laminas guidelines, I'll of course change the code accordingly.

On a side note: Personally, I even think that the whole mocking stuff should be refactored out of PHPUnit, just like Prophecy.

froschdesign replied:
> I almost never use mocks because of the headache they caused…

1. The dependency is the container interface, not the service manager.
2. There is no _overmocking_ in this case.

```php
$container->has('config')
```

https://github.com/laminas/laminas-view/pull/59/files#diff-e6d323bcbed448c76f7d30b10217611dc0ffd9ab40ae27272dd9fe17d87da035R31

```php
$container->get('config')
```
https://github.com/laminas/laminas-view/pull/59/files#diff-e6d323bcbed448c76f7d30b10217611dc0ffd9ab40ae27272dd9fe17d87da035R34

> If the ServiceManager or Container changes, the test should break because the Factory won't work in a real application anymore.

Create a mock object for the interface of the container then it does not matter if the service manager or the container changes.

> Personally, I even think that the whole mocking stuff should be refactored out of PHPUnit, just like Prophecy.

This is the wrong issue tracker! 😉

rieschl replied:
done! I changed the test to use a Mock object

froschdesign reviewed (changes requested):
Looks good so far! 👍 

I have just two small comments.

boesing commented on test/Helper/Service/DoctypeFactoryTest.php:
```diff
@@ -0,0 +1,69 @@
+     * @param array<string, mixed> $config
+     * @return ContainerInterface
+     */
+    private function createContainer(array $config = [])
```
> I almost never use mocks

That is your personal decision and you can do this in your projects. Thats totally fine if you are doing this.
However, in almost every laminas package, we are using mocks instead of concrete implementations as it solves multiple issues:
1. If a bug rises in `laminas-servicemanager`, this package is not affected directly
2. "Unit tests" are supposed to test a specific unit and not how they interact with their dependencies. Since there is no concrete dependency to the `servicemanager`, there is no need to use the servicemanager as a dependency
3. You are explicitly testing the behavior on how the unit is interacting with the container, without using a mock, you cannot verify if the config was called from the container of if it was assembled within the method you are calling

So, please change this to a mock as it would provide better maintainability of this package. 👍

rieschl replied:
Thanks for the input! I changed the test to use a Mock. But I still don't really get why a Mock is better than a "real" object, but that's not something to be discussed here 🙂 Maybe sometime in Slack with a (virtual) beer 😉

rieschl committed:
docs: fix bug in Doctype laminas-mvc documentation

doctype setting in view_manager configuration has to be upper case, use the constant to ensure that

Signed-off-by: Thomas Rieschl <thomas@trinet.at>

rieschl commented:
Okay, I cleaned up the commit history a little bit. Is there anything for me to do?

edit: I added a separate commit at the end which adds the COMPOSER_ROOT_VERSION env to travis. When it's merged, the latest commit can be dumped, according to https://github.com/laminas/laminas-view/pull/58#issuecomment-714539267 . Is that ok, @boesing ?

edit2: do you want me to add an entry in CHANGELOG or is that done when merging the PR?

rieschl committed:
ci: use stable PHP 8.0 version

Signed-off-by: Thomas Rieschl <thomas@trinet.at>

boesing commented:
@rieschl Mostly, maintainers are adding CHANGELOG as its quite hard to determine if a CHANGELOG entry is necessary or not.

e.g. when fixing documentation, there is no need for a dedicated entry within the changelog. there might be other changes (such as inline-phpdoc stuff, e.g.) which does not either.

So its absolutely fine when you omit the CHANGELOG and let maintainers do so 👍🏼 

I'll prepare a release now.

boesing committed:
qa: remove unnecessary `FactoryInterface` implementation and mark factory as `final`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing commented:
@froschdesign @Xerkus Should we think about dropping `view_manager` config in `laminas-mvc` in favor of the `view_helper_config` key?
Thinking about some kind of forward compatibility to enable users to prepare for MVC v4.

boesing committed:
qa: add `CHANGELOG` entry

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: bump `laminas/laminas-servicemanager` dependency

The new `DoctypeFactory` is a generic factory which just needs a PSR-11 container. Thus, we have to bump `laminas/laminas-servicemanager` to v3.3.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing commented:
Thanks, @rieschl!
